File: java/com/masaniwa/negativecamera/MainActivity.java
package com.masaniwa.negativecamera;

import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.view.WindowManager;
import android.widget.Toast;

import org.opencv.android.BaseLoaderCallback;
import org.opencv.android.CameraBridgeViewBase;
import org.opencv.android.LoaderCallbackInterface;
import org.opencv.android.OpenCVLoader;
import org.opencv.core.Mat;

import java.io.IOException;

import static org.opencv.android.CameraBridgeViewBase.CvCameraViewFrame;
import static org.opencv.android.CameraBridgeViewBase.CvCameraViewListener2;

public final class MainActivity extends AppCompatActivity implements CvCameraViewListener2
{
    private static final int quality = 100;
    private static final String directory = "/NegativeCamera/";
    private static final String format = "yyyyMMdd.hhmmss";

    private BaseLoaderCallback loaderCallback = new BaseLoaderCallback(this)
    {
        @Override
        public void onManagerConnected(final int status)
        {
            switch(status)
            {
                case LoaderCallbackInterface.SUCCESS:
                    cameraView.enableView();
                    break;

                default:
                    super.onManagerConnected(status);
                    break;
            }
        }
    };

    private NegativeCamera negativeCamera;
    private CameraBridgeViewBase cameraView;

    private void save()
    {
        final StringBuffer text = new StringBuffer();

        try
        {
            negativeCamera.Save();
            text.append(getString(R.string.saving_success));
        }
        catch(IOException e)
        {
            text.append(getString(R.string.saving_failed));
        }
        finally
        {
            Toast.makeText(this, text, Toast.LENGTH_LONG).show();
        }
    }

    @Override
    protected void onCreate(final Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);

        getWindow().addFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN);
        setContentView(R.layout.activity_main);

        negativeCamera = new NegativeCamera(getContentResolver(), quality, directory, format);

        cameraView = (CameraBridgeViewBase) findViewById(R.id.camera_view);
        cameraView.setCvCameraViewListener(this);

        final FloatingActionButton saveButton = (FloatingActionButton) findViewById(R.id.save_button);
        saveButton.setOnClickListener(new View.OnClickListener()
        {
            @Override
            public void onClick(final View view)
            {
                save();
            }
        });
    }

    @Override
    protected void onResume()
    {
        super.onResume();
        OpenCVLoader.initAsync(OpenCVLoader.OPENCV_VERSION_3_1_0, this, loaderCallback);
    }

    @Override
    public void onPause()
    {
        if(cameraView != null)
        {
            cameraView.disableView();
        }

        super.onPause();
    }

    @Override
    public void onDestroy()
    {
        super.onDestroy();

        if(cameraView != null)
        {
            cameraView.disableView();
        }
    }

    @Override
    public void onCameraViewStarted(final int width, final int height)
    {
        negativeCamera.Initialize(width, height);
    }

    @Override
    public void onCameraViewStopped()
    {
        negativeCamera.Release();
    }

    @Override
    public Mat onCameraFrame(final CvCameraViewFrame frame)
    {
        negativeCamera.Input(frame.rgba());

        return negativeCamera.Output();
    }
}

File: app/src/main/kotlin/com/masaniwa/negativecamera/MainActivity.kt
package com.masaniwa.negativecamera

import android.graphics.Bitmap
import android.graphics.Bitmap.Config
import android.graphics.Bitmap.createBitmap
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.WindowManager.LayoutParams
import android.widget.Toast
import android.widget.Toast.makeText
import kotlinx.android.synthetic.main.activity_main.camera_view
import kotlinx.android.synthetic.main.activity_main.negative_switch
import kotlinx.android.synthetic.main.activity_main.save_button
import org.opencv.android.BaseLoaderCallback
import org.opencv.android.CameraBridgeViewBase.CvCameraViewFrame
import org.opencv.android.CameraBridgeViewBase.CvCameraViewListener2
import org.opencv.android.LoaderCallbackInterface
import org.opencv.android.OpenCVLoader
import org.opencv.android.OpenCVLoader.initAsync
import org.opencv.android.Utils.matToBitmap
import org.opencv.core.Core.bitwise_not
import org.opencv.core.CvType
import org.opencv.core.Mat
import java.io.IOException

/** アプリケーションのメインとなるアクティビティ。 */
class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        window.addFlags(LayoutParams.FLAG_FULLSCREEN)

        setContentView(R.layout.activity_main)

        camera_view.setCvCameraViewListener(cameraListener)

        save_button.setOnClickListener { save() }
    }

    override fun onResume() {
        super.onResume()

        if (!initAsync(OpenCVLoader.OPENCV_VERSION_3_2_0, this, loaderCallback)) {
            notice(R.string.loading_failed)
        }
    }

    override fun onPause() {
        camera_view.disableView()

        super.onPause()
    }

    override fun onDestroy() {
        super.onDestroy()

        camera_view.disableView()
    }

    /**
     * トーストでテキストを表示する。
     * @param id 表示するテキストのリソースID。
     */
    private fun notice(id: Int) {
        makeText(this, getString(id), Toast.LENGTH_LONG).show()
    }

    /**
     * ネガポジ反転した画像をストレージに保存する。
     */
    private fun save() {
        try {
            savePicture(takeBitmap(camera), directory, contentResolver)

            notice(R.string.saving_success)
        } catch (e: IOException) {
            notice(R.string.saving_failed)
        }
    }

    /** OpenCVのライブラリが読み込まれたときのコールバック関数。 */
    private val loaderCallback = object : BaseLoaderCallback(this) {
        override fun onManagerConnected(status: Int) {
            when (status) {
                LoaderCallbackInterface.SUCCESS -> camera_view.enableView()
                else -> super.onManagerConnected(status)
            }
        }
    }

    /** カメラビューのイベントリスナ。 */
    private val cameraListener = object : CvCameraViewListener2 {
        override fun onCameraViewStarted(width: Int, height: Int) {
            camera = Mat(height, width, CvType.CV_8UC3)
        }

        override fun onCameraViewStopped() {
            camera.release()
        }

        override fun onCameraFrame(frame: CvCameraViewFrame): Mat {
            if (negative_switch.isChecked()) {
                bitwise_not(frame.rgba(), camera)
            } else {
                frame.rgba().copyTo(camera)
            }

            return camera
        }
    }

    /** カメラの画像。 */
    private lateinit var camera: Mat
}

/**
 * MatからBitmapを生成する。
 * @param  frame Bitmapの生成元となるMat。
 * @return 生成したBitmap。
 */
private fun takeBitmap(frame: Mat): Bitmap {
    val bitmap = createBitmap(frame.width(), frame.height(), Config.ARGB_8888)

    matToBitmap(frame, bitmap)

    return bitmap
}

/** 画像を保存するディレクトリ。 */
private val directory = "/NegativeCamera/"
